File: app/handlers/payments.py
from aiogram import Router, F
from aiogram.types import Message

from app.database import get_db
from app.services.payments import create_payment
from app.keyboards.payments import coins_keyboard
from app.keyboards.main import main_keyboard

router = Router()

PACKAGES = {
    "💰 50 سکه - 25,000 تومان": (25000, 50),
    "💰 120 سکه - 50,000 تومان": (50000, 120),
    "💰 300 سکه - 100,000 تومان": (100000, 300),
}


# ================== Show coin purchase menu ==================
@router.message(F.text == "💳 خرید سکه")
async def show_coin_packages(message: Message):
    """نمایش پکیج‌های خرید سکه"""
    await message.answer(
        "💰 <b>خرید سکه</b>\n\n"
        "لطفاً یکی از پکیج‌های زیر را انتخاب کنید:",
        parse_mode="HTML",
        reply_markup=coins_keyboard
    )


# ================== Return to main menu ==================
@router.message(F.text == "🔙 بازگشت")
async def back_to_main_menu(message: Message):
    """بازگشت به منوی اصلی"""
    await message.answer(
        "🏠 منوی اصلی:",
        reply_markup=main_keyboard
    )


# ================== Start payment ==================
@router.message(F.text.in_(PACKAGES.keys()))
async def start_payment(message: Message):
    """شروع فرآیند پرداخت"""
    amount, coins = PACKAGES[message.text]

    authority, pay_url = await create_payment(
        amount=amount,
        description=f"خرید {coins} سکه"
    )

    if not authority:
        await message.answer(
            "❌ خطا در اتصال به درگاه پرداخت.\n"
            "لطفاً دوباره تلاش کنید.",
            reply_markup=coins_keyboard
        )
        return

    # Store payment information in the database
    try:
        db = await get_db()
        await db.execute(
            """
            INSERT INTO payments (user_id, amount, coins, authority)
            VALUES (?, ?, ?, ?)
            """,
            (message.from_user.id, amount, coins, authority)
        )
        await db.commit()
        await db.close()

        await message.answer(
            f"💳 <b>اطلاعات پرداخت:</b>\n\n"
            f"💰 مبلغ: {amount:,} تومان\n"
            f"🎁 دریافت: {coins} سکه\n\n"
            f"🔗 برای پرداخت روی لینک زیر کلیک کنید:\n"
            f"{pay_url}\n\n"
            f"⚠️ پس از پرداخت، سکه‌ها به حساب شما اضافه می‌شود.",
            parse_mode="HTML",
            reply_markup=main_keyboard
        )
    except Exception as e:
        print(f"خطا در ذخیره اطلاعات پرداخت: {e}")
        await message.answer(
            "❌ خطا در ثبت اطلاعات پرداخت.",
            reply_markup=coins_keyboard
        )

File: src/handlers/payments.ts
import { Env, TelegramMessage, TelegramCallbackQuery } from '../types';
import { TelegramBot } from '../telegram';
import * as db from '../database';
import * as keyboards from '../keyboards';

// ========================
// Coins Menu Handler
// ========================

export async function handleCoinsMenu(env: Env, bot: TelegramBot, message: TelegramMessage): Promise<void> {
  const userId = message.from.id;
  const user = await db.getUser(env, userId);

  if (!user) {
    await bot.sendMessage(userId, '❌ کاربر یافت نشد.');
    return;
  }

  const text = `
💰 <b>خرید سکه</b>

موجودی فعلی: <b>${user.coins} سکه</b>

با خرید سکه می‌توانید:
• با کاربران جدید چت کنید (هر چت 2 سکه)
• از امکانات ویژه استفاده کنید

یکی از بسته‌های زیر را انتخاب کنید:
  `;

  await bot.sendMessage(userId, text, {
    parse_mode: 'HTML',
    reply_markup: keyboards.coinsKeyboard(),
  });
}

// ========================
// Buy Coins Callback Handler
// ========================

export async function handleBuyCoins(env: Env, bot: TelegramBot, query: TelegramCallbackQuery): Promise<void> {
  const userId = query.from.id;
  const data = query.data;

  if (!data) return;

  const packages: Record<string, { coins: number; amount: number }> = {
    buy_50: { coins: 50, amount: 10000 },
    buy_100: { coins: 100, amount: 18000 },
    buy_200: { coins: 200, amount: 35000 },
    buy_500: { coins: 500, amount: 80000 },
  };

  const pkg = packages[data];

  if (!pkg) {
    await bot.answerCallbackQuery(query.id, 'بسته نامعتبر');
    return;
  }

  // Create payment
  const authority = generateAuthority();
  await db.createPayment(env, userId, pkg.amount, pkg.coins, authority);

  // Create Zarinpal payment
  const paymentUrl = await createZarinpalPayment(env, pkg.amount, authority);

  if (!paymentUrl) {
    await bot.answerCallbackQuery(query.id, 'خطا در ایجاد پرداخت', true);
    return;
  }

  await bot.answerCallbackQuery(query.id);
  
  await bot.sendMessage(
    userId,
    `💳 برای خرید ${pkg.coins} سکه به مبلغ ${pkg.amount.toLocaleString()} تومان، روی لینک زیر کلیک کنید:\n\n${paymentUrl}`,
    { reply_markup: keyboards.mainKeyboard() }
  );
}

// ========================
// Payment Verification Handler
// ========================

export async function handlePaymentCallback(env: Env, authority: string, status: string): Promise<Response> {
  if (status !== 'OK') {
    return new Response('پرداخت ناموفق بود', {
      status: 200,
      headers: { 'Content-Type': 'text/html; charset=utf-8' },
    });
  }

  const payment = await db.getPayment(env, authority);

  if (!payment) {
    return new Response('پرداخت قبلاً بررسی شده', {
      status: 200,
      headers: { 'Content-Type': 'text/html; charset=utf-8' },
    });
  }

  if (payment.status !== 'pending') {
    return new Response('پرداخت قبلاً بررسی شده', {
      status: 200,
      headers: { 'Content-Type': 'text/html; charset=utf-8' },
    });
  }

  // Verify payment with Zarinpal
  const verified = await verifyZarinpalPayment(env, authority, payment.amount);

  if (!verified) {
    await db.updatePaymentStatus(env, authority, 'failed');
    return new Response('پرداخت تایید نشد', {
      status: 200,
      headers: { 'Content-Type': 'text/html; charset=utf-8' },
    });
  }

  // Payment successful
  await db.updatePaymentStatus(env, authority, 'success');
  await db.addCoins(env, payment.user_id, payment.coins);

  // Notify user
  const bot = new TelegramBot(env.BOT_TOKEN);
  try {
    await bot.sendMessage(
      payment.user_id,
      `✅ پرداخت شما با موفقیت انجام شد!\n\n💰 ${payment.coins} سکه به حساب شما اضافه شد.`
    );
  } catch (e) {
    console.error('Failed to notify user:', e);
  }

  return new Response('پرداخت با موفقیت انجام شد!', {
    status: 200,
    headers: { 'Content-Type': 'text/html; charset=utf-8' },
  });
}

// ========================
// Helper Functions
// ========================

function generateAuthority(): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';
  for (let i = 0; i < 36; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

async function createZarinpalPayment(env: Env, amount: number, authority: string): Promise<string | null> {
  const data = {
    merchant_id: env.ZARINPAL_MERCHANT_ID,
    amount: amount,
    description: 'خرید سکه ربات تلگرام',
    callback_url: `${env.CALLBACK_URL}/payment/callback`,
    metadata: {
      authority: authority,
    },
  };

  try {
    const response = await fetch('https://api.zarinpal.com/pg/v4/payment/request.json', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });

    const result = await response.json();

    if (result?.data?.code === 100) {
      const authorityFromZP = result.data.authority;
      return `https://www.zarinpal.com/pg/StartPay/${authorityFromZP}`;
    }
  } catch (e) {
    console.error('Zarinpal payment creation failed:', e);
  }

  return null;
}

async function verifyZarinpalPayment(env: Env, authority: string, amount: number): Promise<boolean> {
  const data = {
    merchant_id: env.ZARINPAL_MERCHANT_ID,
    authority: authority,
    amount: amount,
  };

  try {
    const response = await fetch('https://api.zarinpal.com/pg/v4/payment/verify.json', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });

    const result = await response.json();

    if (result?.data?.code === 100 || result?.data?.code === 101) {
      return true;
    }
  } catch (e) {
    console.error('Zarinpal payment verification failed:', e);
  }

  return false;
}
